components/sup/src/manager/mod.rs

```diff
@@ -778,37 +867,107 @@ impl Manager {
                 next_cpu_measurement = SteadyTime::now() + TimeDuration::seconds(1);
                 cpu_start = ProcessTime::now();
             }
+        }; // end main loop
+
+        // When we make it down here, we've broken out of the main
+        // Supervisor loop, which means it's time to shut down. Based
+        // on the value we broke out of the loop with, we may need to
+        // shut services down. We do that out here, so we can run the
+        // shutdown futures directly on the reactor, and ensure
+        // they're all driven to completion before we exit.
+
+        // Stop the ctl gateway; this way we'll stop responding to
+        // user commands as we're trying to shut down.
+        ctl_shutdown_tx.send(()).ok();
+
+        match shutdown_mode {
+            ShutdownMode::Restarting => {}
+            ShutdownMode::Normal | ShutdownMode::Departed => {
+                outputln!("Gracefully departing from butterfly network.");
+                self.butterfly.set_departed();
+
+                let mut svcs = self
+                    .state
+                    .services
+                    .write()
+                    .expect("Services lock is poisoned!");
+
+                for (_ident, svc) in svcs.drain() {
+                    runtime.spawn(self.stop(svc));
+                }
+            }
+        };
+
+        // Allow all existing futures to run to completion.
+        runtime
+            .shutdown_on_idle()
+            .wait()
+            .expect("Error waiting on Tokio runtime to shutdown");
+
+        release_process_lock(&self.fs_cfg);
+        self.butterfly.persist_data();
+
+        match shutdown_mode {
+            ShutdownMode::Normal | ShutdownMode::Restarting => Ok(()),
+            ShutdownMode::Departed => Err(sup_error!(Error::Departed)),
```

Comment:
I'm a little confused.  Here, we throw an error if `shutdown_mode` is `Departed`, but up above in the match statement, we output `Gracefully departing` for that same `shutdown_mode`.  Those two things seem to contradict each other.  Is departure normal or an error?